src/openforms/submissions/tests/form_logic/test_binding.py

```diff
@@ -0,0 +1,375 @@
+            service_fetch_configuration=ServiceFetchConfigurationFactory.build(
+                service=self.service,
+                path="get",
+            )
+        )
+
+        value = bind(var, {})
+
+        self.assertEqual(value["url"], "https://httpbin.org/get")
+
+    @given(data_mapping_values())
+    @example("../otherendpoint")
+    def test_it_can_construct_simple_path_parameters(self, field_value):
+        # https://swagger.io/docs/specification/describing-parameters/#path-parameters
+        context = {"seconds": field_value}
+
+        var = FormVariableFactory.build(
+            service_fetch_configuration=ServiceFetchConfigurationFactory.build(
+                service=self.service,
+                path="delay/{seconds}",  # this is not defined as number in the OAS
+            )
+        )
+
+        with requests_mock.Mocker() as m:
+            m.get(requests_mock.ANY)
+            _ = bind(var, context)
+            request = m.last_request
+
+        head, rest = request.url[:26], request.url[26:]
+        self.assertEqual(head, "https://httpbin.org/delay/")
+        # the service should get the value
+        self.assertEqual(unquote(rest), str(field_value))
+
+        # I guess if the rest is unmangled, the rest of the assertions are not
+        # really needed; I think
+        # ∄ a way for Bobby Tables to drop a table ∧ become an unmangled entry in it
+        # but asserting is cheaper than a CVE
+
+        # it shouldn't change other parts of the request
+        self.assertIs(request.body, None)
+        self.assertIs(request.query, "")
+        self.assertEqual(set(request.headers), DEFAULT_REQUEST_HEADERS)
+        # and a weak assertion
+        self.assertNotIn(str(field_value), request.headers.values())
+
+    @skip("TODO")
+    @given(
+        st.one_of(
+            st.lists(elements=data_mapping_values()),
+            st.dictionaries(
+                keys=st.text(), values=st.one_of(st.text(), st.integers(), st.floats())
+            ),
+        )
+    )
+    def test_it_can_construct_product_type_path_parameters(self, field_value):
+        # serialize lists and dicts of values as specified by `style` and
+        # `explode` in the oas-spec of the service
+        # https://swagger.io/docs/specification/serialization/
+        ...
+
+    @given(
+        field_value=data_mapping_values(),
+        question_mark=st.one_of(st.just("?"), st.just("")),  # optional
+    )
+    def test_it_can_construct_simple_query_parameters(self, field_value, question_mark):
+        # https://swagger.io/docs/specification/describing-parameters/#query-parameters
+        context = {"some_field": field_value}
+
+        var = FormVariableFactory.build(
+            service_fetch_configuration=ServiceFetchConfigurationFactory.build(
+                service=self.service,
+                path="response-headers",
+                query_params=question_mark + "freeform={some_field}",
```

Comment:
I'm really picturing a UX where users enter key/values and can enter lists as values too, which get serialized down to a mapping/dict in the backend rather than strings. So the question mark in that case becomes irrelevant, especially if we feed the (evaluated/templated out) dict values to something like `furl` or low-level encoding which takes care of all the peculiarities. IIRC the ZDS client can also take params as a dict which is then serialized properly by requests.

Reply:
I think refactoring the Mapping type values a bit tighter maybe to an [ArrayField](https://docs.djangoproject.com/en/3.2/ref/contrib/postgres/fields/#arrayfield) is also an option... 
If it's okay with you, I'd like to hot potato this decision over to #2694

Reply:
I've instructed Steven to make a key-value UX in the modal for this, where the values may be arrays (especially for querystrings, but only strings for headers). ArrayField cannot store arrays afaik and HStoreField also only takes strings as values, so we'll have to stick to JSONField.

Okay to move this into a different ticket :+1: